test/tests/functional/pbs_test_tpp.py

```diff
@@ -386,30 +429,414 @@ def test_multiple_comm_with_mom(self):
+        """
+        self.common_setup()
+        hostA_ip = socket.gethostbyname(self.hostA)
+        hostC_ip = socket.gethostbyname(self.hostC)
+        a = {'PBS_COMM_ROUTERS': self.hostA}
+        self.set_pbs_conf(host_name=self.hostC, conf_param=a)
+        leaf_val = self.hostA + ":17001" + "," + self.hostC
+        b = {'PBS_LEAF_ROUTERS': leaf_val}
+        self.set_pbs_conf(host_name=self.hostA, conf_param=b)
+        leaf_val = hostC_ip + "," + self.hostA + ":17001"
+        b = {'PBS_LEAF_ROUTERS': leaf_val}
+        self.set_pbs_conf(host_name=self.hostB, conf_param=b)
+        self.common_steps_for_comm_failover()
+
+    def copy_pbs_conf_to_non_default_path(self):
+        """
+        This function copies the pbs.conf from default location
+        to non default location
+        """
+        # Retrieve temporary directory
+        tmp_dir = self.du.get_tempdir(hostname=self.server.hostname)
+        msg = "Unable to get temp_dir"
+        self.assertNotEqual(tmp_dir, None, msg)
+        self.new_conf_path = os.path.join(tmp_dir, "pbs.conf")
+
+        # Copy pbs.conf file to temporary location
+        rc = self.du.run_copy(src=self.pbs_conf_path, dest=self.new_conf_path)
+        msg = "Cannot copy %s " % self.pbs_conf_path
+        msg += "%s, error: %s" % (self.new_conf_path, rc['err'])
+        self.assertEqual(rc['rc'], 0, msg)
+
+        # Set the PBS_CONF_FILE variable to the temp location
+        os.environ['PBS_CONF_FILE'] = self.new_conf_path
+        self.logger.info("Successfully exported PBS_CONF_FILE variable")
+
+        self.server.pi.conf_file = self.new_conf_path
+        self.pbs_restart(self.server.hostname)
+        self.logger.info("PBS services started successfully")
+
+    @requirements(num_moms=2, num_comms=2)
+    def test_comm_failover_with_nondefault_pbs_conf(self):
+        """
+        This test verifies communication between server-mom and
+        between mom when multiple pbs_comm are present in cluster
+        with pbs_comm failover when PBS_LEAF_ROUTERS has ipaddress,
+        port number and hostname as values and pbs.conf is in
+        non default location
+        Configuration:
+        Node 1 : Server, Sched, Mom, Comm
+        Node 2 : Mom
+        Node 3 : Comm
+        """
+        self.common_setup()
+        hostA_ip = socket.gethostbyname(self.hostA)
+        hostC_ip = socket.gethostbyname(self.hostC)
+        a = {'PBS_COMM_ROUTERS': self.hostA}
+        self.set_pbs_conf(host_name=self.hostC, conf_param=a)
+        leaf_val = self.hostA + ":17001" + "," + self.hostC
+        b = {'PBS_LEAF_ROUTERS': leaf_val}
+        self.set_pbs_conf(host_name=self.hostA, conf_param=b)
+        leaf_val = hostC_ip + "," + self.hostA + ":17001"
+        b = {'PBS_LEAF_ROUTERS': leaf_val}
+        self.set_pbs_conf(host_name=self.hostB, conf_param=b)
+        self.copy_pbs_conf_to_non_default_path()
+        self.common_steps_for_comm_failover()
+
+    @requirements(num_moms=3, num_comms=3)
```

Comment:
Does the requirement decorator know this test needs 6 nodes?  In this case num_moms+num_comms = 6.  The above test needs 3 nodes, but num_moms+num_comms = 4.

Reply:
By default, requirement decorator has all values set to 1(num_mom=1, num_comm=1, num_server=1). 
In the above test, I would need 2 moms and 2 comms out of which 1 mom and 1 comm will be on server by default(host1) and on the rest of the hosts. second mom is present on host2 and comm is present on host3

Reply:
I wasn't exactly clear with my question.  The @requirements() decorator will skip the test if the requirements aren't met.  How does the decorator know how you are spreading out your moms and comms in order to skip the test?

Reply:
Value in the requirement decorator is number of daemons and not number of nodes. Along with no of deamons, there are three flags in decorator which are no_comm_on_server,
no_mom_on_server and no_comm_on_mom. All this info helps us know about the cluster through the daemons. The value passed in -p moms and comms are set on these daemons objects. Currently in my tests I am creating the list of node names by accessing these daemon objects.

Reply:
I'm not all that familiar with the requirement decorator.  I am just wanting to make sure these tests are skipped properly if the requirements aren't met.  You have a complex node setup.  It sounds like you are confident that it will be, so I am OK with that.

Reply:
you are running the tests separately by providing test names to pbs_benchpress.
Wasn't the purpose of requirements decorator to skip the tests if requirements are not met?

Reply:
Each of these tests have different combinations of cluster daemons. So to run all the tests if I give the whole suite name, Only few of them will run and rest of the tests get skipped. To avoid that, I am running tests by providing the test names explicitly so that I will cover execution of all the tests.

Reply:
okay - thanks for explaining.